[stm32f1] More robust workaround for I2C errata

vvuk opened the pull request:
The stm32f1 has an issue that can cause the i2c Lines to be "stuck". This patch implements a more thorough version of the workaround from the errata. Klipper currently only applies the errata during `i2c_setup`, but the condition can happen at any point. This adds detection of the condition (via bus busy) and tries to apply the errata process again in i2c_wait.

The original patch comes from Sovol, though I can't link to it because it lives in the .git directory that they ship on the eMMC with the Sovol Zero :p I've refactored and cleaned up much of the patch, not sure how to give attribution though. Without this patch, the i2c devices on the Sovol Zero (and presumably SV08 Max, I think it's the same one) toolhead are not usable.

Though reading through this again... this will double-apply an i2c_wait timeout in i2c_start in case of bus busy on _all_ stm32s, not just those that need to have the errata applied. I'll fix that. I'm not deeply familiar with the stm32 though, so there may be details that I've missed.

vvuk commented:
Er, I also just noticed a logic error with how error codes are handled (as bits or as values)... that was present in the original patch, and I think it "accidentally" worked. Will fix.

nefelim4ag commented on src/stm32/i2c.c:
```diff
@@ -51,18 +51,69 @@ static const struct i2c_info i2c_bus[] = {
+i2c_us_delay(uint32_t us)
+{
+    uint32_t t = timer_read_time() + timer_from_us(us);
+    while(t > timer_read_time());
```
This is generally not correct because timer could and will overflow.
Please use `timer_is_before()`:
```
    while (timer_is_before(timer_read_time(), end))
        ;
```

vvuk replied:
Oops, yes. I've actually written this function a few times now in various places. Feels like I should just add a `spin_us(usec)` function somewhere?

nefelim4ag replied:
Sometimes, I also wish for it.
But it so happened historically that local pauses are made locally.

vvuk commented on src/stm32/i2c.c:
```diff
@@ -51,18 +51,69 @@ static const struct i2c_info i2c_bus[] = {
+    const struct i2c_info *ii = container_of((I2C_TypeDef * const *)i2c, struct i2c_info, i2c);
+    uint32_t val;
+    (void)val;
+    val = i2c->SR1;
```
Yeah, this was in the original patch.. I didn't look at the assembly output, so wasn't sure if there was anything magic. I'll remove the variable.

vvuk commented:
Fixed a number of things, but now untested. I'll test ASAP but might be a day or two.

nefelim4ag commented on src/stm32/i2c.c:
```diff
@@ -51,18 +51,70 @@ static const struct i2c_info i2c_bus[] = {
+    uint32_t scl_bit = GPIO2BIT(ii->scl_pin);
+    GPIO_TypeDef *sda_regs = digital_regs[GPIO2PORT(ii->sda_pin)];
+    uint32_t sda_bit = GPIO2BIT(ii->sda_pin);
+    #define SCL_ODR_SET(v) do { scl_regs->BSRR = (v) ? (scl_bit<<16) : scl_bit; } while (0)
```
Seems to be reversed.
Generally, here is gpio_write()

```
// src/stm32/gpio.c
void
gpio_out_write(struct gpio_out g, uint32_t val)
{
    GPIO_TypeDef *regs = g.regs;
    if (val)
        regs->BSRR = g.bit;
    else
        regs->BSRR = g.bit << 16;
}
```

vvuk replied:
Ugh, quite. (Thank you for the in-depth review here!) Is there some way that I can just use `gpio_out_write`? Maybe I should synthesize a `struct gpio_out` for the scl/sda pins?

nefelim4ag commented:
Overral, the problem you are describing sounds bad.
(Maybe just let's use SW I2C? There is no errata :D).

Sorry, there would be a lot of comments.

High-level comments not in particular order:
* I suggest splitting the part where you are adding the error handling (like the first handling NACK, and the second actual duct tape) (maybe even a separate PR, it is much more easily mergeable).
* JFYI: Error codes are supposed to give some information to the user/code above. Like START NACK, it would mean there is no device on the line or it is not responding. Maybe it would allow us to handle them somehow in the future (generally, it would be a retry).
* I would suggest avoiding mixing the device-specific/errata codes with the general ones.


Mmm... Further thinking, well, does it really fix the issue?
Seems like it adds a lot of code to just throw an error and crush everything.
Ahh, there is no BUSY error handling in the above call stack, so no shutdown, how convenient.

I suspect the real question here is how we want to work around that, if, in any place in the transaction, the I2C bus could just technically die.
So, the data/transaction is garbled.
So, the current PR, with proper implementation, still, should crash the MCU.

It is possible to split the F1 I2C to a separate file, making it easier to do `goto out:`, which could simplify the code.
It is possible to internally retry the transaction in specific circumstances. I think.

But those, anyway, would require Kevin's attention and opinion.

---
Does the current approach of reset fail to reset the bus? If so, part of this code is already nice to have.
Because, according to the errata, device reset does not unlock the bus.
![image](https://github.com/user-attachments/assets/54906ee3-cccd-4935-b259-e70616744c8a)

If you have a specific pair of devices that does cause pain, it probably makes more sense to:
1. Implement basic error forwarding.
2. Do the device-specific handling (taking into account your previous work on Eddy).

vvuk commented:
> JFYI: Error codes are supposed to give some information to the user/code above. Like START NACK, it would mean there is no device on the line or it is not responding. Maybe it would allow us to handle them somehow in the future (generally, it would be a retry).

The only added error is "Busy", which seems informative enough? ("START_BUSY" doesn't actually make sense, I've removed it. Or rather, maybe START_BUSY is really the only BUSY...) It would allow higher level code to retry if we ever passed errors up instead of shutting down.

> Mmm... Further thinking, well, does it really fix the issue?
> Seems like it adds a lot of code to just throw an error and crush everything.
There are two core changes:

1. Change the errata workaround routine to more closely follow the description in the errata. I think I match it now, except for the pin checks after the writes to the GPIO regs.
2. At the start of a transaction only, if the bus is seen to be busy and we're on stm32f1, apply the errata workaround and try again. Currently the existing errata workaround is applied only when the i2c bus is initialized the first time.
3. Forward errors more consistently (i.e. currently on stm32 there's an immediate shutdown on NACK or timeout, instead of making it up into the shutdown_on_err or higher level code)

I can split these into separate patches.

> Ahh, there is no BUSY error handling in the above call stack, so no shutdown, how convenient.
That's an oversight; it should definitely be handled in `shutdown_on_err`. I'll add.

> I suspect the real question here is how we want to work around that, if, in any place in the transaction, the I2C bus could just technically die. So, the data/transaction is garbled.

Well -- not at any place in a transaction, just at the start of one in `i2c_start`. If the bus dies sometime in the middle, some kind of error would occur in the i2c_wait following read/write and it would be immediately propagated.

vvuk commented:
> This is just my nonsense. It is possible to internally restart the I2C transaction when it dies on the busy flag.

But this is what it does, except it doesn't restart a transaction from the middle, only at the start. (Well, not technically true, though that can be fixed.)

This function is introduced:

```
static int
i2c_wait_with_retry(I2C_TypeDef *i2c, uint32_t set, uint32_t clear, uint32_t timeout,
                    uint8_t want_retry)
{
    uint8_t did_retry = want_retry ? 0 : 1;
retry:
```

where the wait will retry once (and only once), after applying the workaround, if a busy is seen.

That wait-with-retry is only used in `i2c_start`. It's currently used both to detect the start condition as well as to detect that the address phase is complete; I actually don't think it should be used in the second, and only should be used to detect the start condition. At that point it should be safe to reset the internal i2c machinery of the chip as many times as we want, but after the start during the transaction, any errors are immediately errors.

It's definitely not a 100% fix for the issue, but if the condition can happen at any point, then a 100% fix is not possible without data corruption. But not much we can do about that other than recommend software i2c.

vvuk committed:
[stm32f1] More robust workaround for I2C errata

The stm32f1 has an issue that can cause the i2c Lines
to be "stuck". This patch implements a more thorough
version of the errata fix, and adds some retries
in i2c paths.

Signed-off-by: Vladimir Vukicevic <vladimir@pobox.com>

KevinOConnor commented:
> Which from my reading of the errata ("This issue occurs randomly.") means that the errata condition can be observed at the start of any communication, not just at startup.

For what it is worth, my reading of the errata is a little different.  I read it as saying that there is an "analog filter" on the chip that reports the state of the gpio pins after filtering out electric line noise, and that this "analog filter" may not report the correct state of the gpio pins if it has never observed an edge transition on the gpio pins.  So, it's possible for the "analog filter" to report both gpio pins as "low" when they are actually both in a "high" state.  To clear this issue, one purposely toggles the state of the gpio pins at least once, and then the "analog filter" is in sync with the gpio pins.  Normally this quirk is irrelevant, but if one configures the pins for i2c as a transmitter and the analog filter incorrectly thinks that the two gpio pins are "low" then the transmitter will never change the gpio pin state, the analog filter will thus never get back in sync, and one is stuck forever waiting to transmit.  My read of the spec is that the initial state of the "analog filter" is random as long as it has never seen a transition on the gpio pins, not that the "analog filter" will randomly report incorrect state after it has synced with the gpio pins.

In any case, it's certainly possible I'm misunderstanding something.

Cheers,
-Kevin

vvuk commented:
That reading also makes sense! I am now definitely reproducing the "bus busy" condition right now on the toolboard here, after the errata is applied at startup, so I can do some investigation. It's totally possible that it's being caused by something else though, so I want to dig in. As far as I know the only thing on i2c there is a ldc1612, so nothing else should be trying to become master on the i2c bus.

vvuk commented:
Okay, tested on hardware, and updated the patch. I don't actually understand how sovol's orginal code (at least the source they ship on the printer, I don't know if that's what the firmware is from) could have worked.

I am definitely seeing the BUSY condition after init. In my previous patch (and in sovol's), I waited for the timeout time to expire _before_ checking BUSY, and then triggering the errata reset. That worked. If I immediately returned from the `i2c_wait` loop if BUSY was asserted (i.e. before the timeout), I never made progress.

If I wait for the timeout to expire (or, I would guess, wait some amount of time but I didn't test an arbitrary amount), and I see a BUSY, and I apply the errata, the rest of the transaction finishes fine. (Even though we're already past the timeout, the timeout is only checked in i2c_wait after all the success/exit conditions.) Communication is reliable; I'm testing with repeated bed meshes and I have not had an issue yet.

Sovol's original (and mine) just ham-fistedly applied the errata in i2c_wait and then continued on; this led to an I2C_NAK sometimes because the I2C transaction wouldn't actually have been started. (it'll be a fresh "i2c init").

So, this patch now:
1. Expands out the errata-applying function to (I think) follow more closely what's in the errata
2. Applies the fix _only_ at the start of a transaction, if BUSY is seen after the timeout expires. It's only tries to apply it once. It also properly restarts the transaction (i.e. asserts I2C_CR1_START etc again), so there's no corruption. It reserves a portion of the total timeout for handling this (to avoid the "well, timeout expired, but we're just hitting all success paths anyway" situation).
3. Adds error-code returning paths in the stm32 i2c code, so that errors are sent back up to i2ccmds and can be handled in the regular shutdown-on-err path (instead of i2c_wait doing a shutdown).

I think these are overall safe changes. However, there's a chance the _actual_ issue is just that maybe just in this case, the errata fix is being applied too early. i.e. before the i2c peripheral is configured, which might be what causes all the pieces (i.e. the analog filters) to come into play. Despite reproducing the issue dozens of times in the past hour, I now can't reproduce it to verify this :) Even if that's the case though, the errata does say "This can occur after an MCU power-on reset, _or during ESD stress_", which means that it could happen beyond just startup.

KevinOConnor commented:
Okay, thanks for the information.  However, I suspect there is something else going on beyond the errata.  There are a lot of stm32f103 chips running Klipper and I can't see changing this code (risking a regression to such a large number of currently happy users) without a solid understanding of the underlying problem.

So, I think to make progress on this, we'll need to get more diagnostic information.  Can you provide the Klipper logs of the event when running the pristine Klipper code?  Do you have an oscilloscope or logic analyzer that you could attach to the wires to show what the actual wire state is at the time of the event (eg, https://www.klipper3d.org/CANBUS_Troubleshooting.html#using-a-logic-analyzer-on-the-canbus-wiring )?

Another tool that may help would be to use console.py to dump the memory from the gpio peripheral and i2c peripheral after the event occurs.  For example, run Klipper and initiate the event then stop Klipper and connect with console.py ( https://www.klipper3d.org/Debugging.html#manually-sending-commands-to-the-micro-controller ) and use the DUMP command to show the hardware state (`DUMP  : Dump memory (eg, "DUMP 0x12345678 100 32")`).

-Kevin

nefelim4ag commented:
JFYI, I've received the following log at the channel kipper-development on Discord.
I guess it is related, and this is how the error probably should look.
The MCU is: STM32F103 Blue Pill
```
shutdown clock=70378161 static_string_id=Failed to send i2c addr
```

nefelim4ag commented:
@LupusTheCanine,
This is how you could help:
I think the real error is not the `Unable to obtain 'i2c_read_response' response`
If you get this, that probably mean something is really off, for example - I2C errors disabled in the code.

So, it would help if you can get the real error under the mainline/prestine klipper.
Then there could be a round of debug changes to get the idea of what is going wrong/where on the chip.
And after that there could be changes to workaround the issue.

If you wish, you could contact me on the klipper discord.

About this specific patchset, right now it should be rebased at least, and probably would require same information to move it somewhere, because in the current state of the mainline code is based on the ERRATA, which might be ambition in a sense or incomplete.

Also, as mentioned above, you could use the SW I2C implementation, and if there are issues, klippy.log could help.

Hope that clarifies something.

LupusTheCanine commented:
@nefelim4ag 

> Software probably returns zeroed responses because either rate is too high (400k) and there are no pull-ups or errors has been disabled.

There are ~4.7k pull-ups on both I2C lanes on the probe side.

Here is software serial test with 100k rate.
[klippy (3).log.txt](https://github.com/user-attachments/files/24155896/klippy.3.log.txt)

What do you mean by "prestine"?

I am on mainline with eddy-ng installed though AFAIK it doesn't change I2C handling.
